Pixel retrying

samsymons opened the pull request:
<!--
Note: This checklist is a reminder of our shared engineering expectations. Feel free to change it, although assigning a GitHub reviewer and the items in bold are required.

⚠️ If you're an external contributor, please file an issue first before working on a PR, as we can't guarantee that we will accept your changes if they haven't been discussed ahead of time. Thanks!
-->

Task/Issue URL: https://app.asana.com/0/72649045549333/1208300963176861/f
Tech Design URL:
CC:

**Description**:

This PR adds pixel retrying behavior for a small set of pixels.

<!--
If at any point it isn't actively being worked on/ready for review/otherwise moving forward strongly consider closing it (or not opening it in the first place). If you decide not to close it, use Draft PR while work is still in progress or use `DO NOT MERGE` label to clarify the PRs state and comment with more information.
-->

**Steps to test this PR**:

1. To test this, we should either fire real pixels by disabling `dryRun` mode or simulate a failed error call by adding an error to `Pixel.swift#240` - adjust the following steps based on which path you take (i.e., enabling Airplane Mode to test real pixel call failures, or removing the fake error and rebuilding)
1. Sign into a Privacy Pro account, easiest way to get one is [here](https://app.asana.com/0/0/1208034056804149); once you have it then open the Privacy Pro settings and follow the "I have an account" flow
2. Enable and start the VPN by opening the VPN UI
3. Check that you see log entries that say "Saving persistent pixel" followed by `m_netp_controller_start_attempt` and `m_netp_controller_start_success` pixels (note: even if you're offline, it's expected to see the success pixel, as the success pixel only reports the request we send to the OS to start the VPN, which is not affected by connectivity)
4. Now put your device back into a state that lets it send pixels (disable Airplane Mode, or remove the hardcoded pixel send error) and wait 60 seconds
5. Background and foreground the app and check that the pixels have been retried; look for `Persistent pixel retrying` in the logs and then search for the pixel names (`m_netp_controller_start_attempt`, `m_netp_controller_start_success `)

<!--
Before submitting a PR, please ensure you have tested the combinations you expect the reviewer to test, then delete configurations you *know* do not need explicit testing.

Using a simulator where a physical device is unavailable is acceptable.
-->

**Definition of Done (Internal Only)**:

* [ ] Does this PR satisfy our [Definition of Done](https://app.asana.com/0/1202500774821704/1207634633537039/f)?

**Copy Testing**:

* [ ] Use of correct apostrophes in new copy, ie `’` rather than `'`

**Orientation Testing**:

* [ ] Portrait
* [ ] Landscape

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone 8
* [ ] iPhone X
* [ ] iPhone 14 Pro
* [ ] iPad

**OS Testing**:

* [ ] iOS 15
* [ ] iOS 16
* [ ] iOS 17

**Theme Testing**:

* [ ] Light theme
* [ ] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

samsymons committed:
Add pixel persistence.

samsymons committed:
Add the PersistentPixel implementation and tests.

samsymons committed:
Update tests.

samsymons committed:
Merge branch 'main' into sam/persistent-pixels

* main:
  Ensure toast closures are called on the main thread (#3347)
  Update survey builder OS version (#3348)
  BSK - Add feature flag for SKAN API (#3356)
  Remove checking for negative attribution case  (#3355)
  C.S.S Bump (Via BSK) (#3346)
  Update Onboarding gradients (#3350)
  Alessandro/onboarding copy and private search options (#3349)
  Onboarding Intro - Add choose address bar position (#3340)

samsymons committed:
Add timestamp and retry parameters.

samsymons committed:
Undo unused changes.

samsymons committed:
Fix indentation.

samsymons committed:
More cleanup.

samsymons committed:
Test the pixel limit.

samsymons committed:
Handle incoming pixels when the queue is being processed.

samsymons committed:
Tidy up tests.

samsymons committed:
Clear out pending pixels when queue processing is complete.

samsymons committed:
Add dispatchPrecondition.

samsymons committed:
Add PersistentPixel logs.

samsymons committed:
Migrate VPN controller pixels to PersistentPixel.

samsymons committed:
Update PersistentPixel tests.

samsymons committed:
Fix broken test.

samsymons committed:
Test the timestamp check.

samsymons committed:
Update PersistentPixel.

samsymons committed:
Surface persistence errors as their own types.

samsymons committed:
Relocate mock.

samsymons committed:
Add PersistentPixel to the PacketTunnelProvider.

samsymons committed:
Merge branch 'main' into sam/persistent-pixels

* main:
  Remove VPN feature flag checks (#3334)
  Add Marketplace Postback handling (#3357)
  SKAD4 crash fix (#3361)
  Enroll all internal users in experiment && Update BSK (#3359)
  update for macOS: visited links (#3353)

samsymons committed:
Filter out pixels older than 28 days.

samsymons committed:
Add additional logs to help debugging.

samsymons commented on DuckDuckGo/NetworkProtectionTunnelController.swift:
```diff
@@ -128,16 +130,33 @@ final class NetworkProtectionTunnelController: TunnelController, TunnelSessionPr
+            error: nil,
+            includedParameters: [.appVersion, .atb],
+            withAdditionalParameters: [:],
+            onComplete: { _ in })
```
Error handling here could be improved by sending a write failure pixel

samsymons committed:
Merge branch 'main' into sam/persistent-pixels

# By Daniel Bernal (12) and others
# Via Daniel Bernal (5) and others
* main: (46 commits)
  Release 7.139.0-4 (#3411)
  Onboarding highlights experiment updates (#3406)
  fix suggestions performance (#3405)
  For third party requests differentiate if they are affiliated with first party (#3386)
  Bump BSK to pull in C-S-S 6.19.0 (#3396)
  Release 7.139.0-3 (#3399)
  Onboarding Highlights Ship Review  (#3380)
  add assertions for tabs in suggestions (#3394)
  Release 7.139.0-2 (#3398)
  Bump BSK to Include C.S.S 6.17 (#3397)
  Bump BSK which includes C.S.S 6.17 (#3395)
  Rever BSK branch
  Revert "Bump C.S.S"
  Bump C.S.S
  translations for Switch to Tab (#3391)
  Remove Favorites section header from NTP (#3381)
  Release 7.139.0-1 (#3389)
  Add origin to /apps URL (#3378)
  chery pick returning user fix
  Release 7.138.1-0 (#3388)
  ...

# Conflicts:
#	DuckDuckGo/AppDependencyProvider.swift

bwaresiak commented on Core/PersistentPixelStoring.swift:
```diff
@@ -0,0 +1,157 @@
+
+    func replaceStoredPixels(with pixels: [PersistentPixelMetadata]) throws {
+        try fileAccessQueue.sync {
+            try writePixelDataToFileSystem(pixels: pixels)
```
Shouldn't we also check the `pixelCountLimit` here?

samsymons replied:
Done; this function has been replaced with new `append` and `remove` functions, where the `append` one can append an array of any size but will take the last 100 elements before writing them back to disk.

bwaresiak commented on Core/PersistentPixel.swift:
```diff
@@ -0,0 +1,326 @@
+        defer { pixelProcessingLock.unlock() }
+
+        for pixel in self.failedPixelsPendingStorage {
+            try? self.persistentPixelStorage.append(pixel: pixel)
```
Can we append as batch?

samsymons replied:
This is resolved now, we no longer re-append failed pixels as the `failedPixelsPendingStorage` mechanism has been entirely removed.

We now collect a set of successfully-retried pixel UUIDs that get filtered out at the end of the retrying process. This means that incoming pixels can now be written to storage while processing is active, since we're no longer replacing the entire post-processing pixel array.

bwaresiak commented on Core/PersistentPixelStoring.swift:
```diff
@@ -0,0 +1,157 @@
+            pixels.append(pixel)
+
+            if pixels.count > pixelCountLimit {
+                pixels.removeFirst()
```
That drop does not necessairly remove the oldest - because of how the failed pixel array is behaving.

samsymons replied:
The order of the array is now preserved - pixels get written immediately upon failure, and we filter them out when they get successfully retried, so we're no longer juggling two arrays at once and combining them in unexpected ways.

That said, it's still possible for pixels to get a little bit out of order; for example:
1. The app tries to send pixel 1, and then tries to send pixel 2 a second later
2. Pixel 1, for whatever reason, takes longer for its network request to fail
3. Because we capture the timestamp at the time of the pixel send initiation, pixel 1 is now positioned _after_ pixel 2 in the storage despite having a timestamp 1 second earlier

This feels low risk to me so I've opted to leave it, but adding a sorting mechanism to the storage call wouldn't be terribly expensive; let me know how you feel.

bwaresiak reviewed (changes requested):
I've went through the code, overall it looks good but there are some aspects I'd like to go through with you, so just ping me when you're around. :)

samsymons committed:
Merge branch 'main' into sam/persistent-pixels

* main:
  Update BSK version to support userTips NetP Flag (#3416)
  Release 7.140.0-0 (#3418)
  Adds new debug screen to view raw credentials data (#3400)
  Bump rollup from 2.79.1 to 3.29.5 (#3385)
  fix UI tests broken by tabs suggestions (#3402)
  Fix ability to edit Synced device name (#3404)
  Fix failing password-management maestro test (#3413)

samsymons committed:
Merge branch 'main' into sam/persistent-pixels

* main:
  Refactor Core Data usage in RemoteMessagingStore (#3425)
  Release 7.140.0-1 (#3428)
  Updates Internal release with 7.139.1 hotfix changes (#3426)
  Disabling autocompletion for website suggestions on macOS (#3401)

samsymons committed:
Update pixels to be removed by UUID, and remove `failedPixelsPendingStorage` usage.

Still to do: update PersistentPixel to use a simple dispatch queue for access control, we shouldn't need the lock any more.

samsymons commented on Core/PersistentPixel.swift:
```diff
@@ -218,46 +217,49 @@ public final class PersistentPixel: PersistentPixelFiring {
                let originalSendDate = dateFormatter.date(from: originalSendDateString),
                let date28DaysAgo = calendar.date(byAdding: .day, value: -28, to: currentDate) {
                 if originalSendDate < date28DaysAgo {
+                    pixelIDsAccessQueue.sync {
+                        _ = pixelIDsToRemove.insert(pixelMetadata.id)
+                    }
```
We were getting outdated pixels removed for free because they weren't getting added to the failed pixels array, so now we need to explicitly return them as part of the `pixelIDsToRemove`.

samsymons committed:
Replace lock and boolean flag with a single queue.

This queue uses a date check to prevent duplicate calls from triggering duplicate processing events. If two events are queued one after the other, the second one will return early because of this check.

samsymons commented on Core/PersistentPixelStoring.swift:
```diff
@@ -0,0 +1,176 @@
+    }
+}
+
+protocol PersistentPixelStoring {
+    func append(pixels: [PersistentPixelMetadata]) throws
+    func remove(pixelsWithIDs: Set<UUID>) throws
+    func storedPixels() throws -> [PersistentPixelMetadata]
+}
```
There's now no way to remove pixels from the array without referencing them by ID, so we don't have to worry about newly-appended pixels getting overwritten somehow, like was possible with the old approach.

samsymons committed:
Merge branch 'main' into sam/persistent-pixels

# By bwaresiak
# Via GitHub
* main:
  Improve unit tests reliability (#3430)

# Conflicts:
#	Core/PixelFiring.swift
#	DuckDuckGoTests/MockPixelFiring.swift

samsymons committed:
Resolve build failures after merge conflicts.

Tests do not yet pass, this will be fixed in the next commit.

samsymons committed:
Fix up main test failures. Still a flaky test failure to fix.

samsymons committed:
Wait for `sendQueuedPixels` to complete its work before proceeding.

samsymons committed:
Fix SwiftLint.

samsymons committed:
Update logger usage.

samsymons committed:
Merge branch 'main' into sam/persistent-pixels

* main:
  Release 7.141.0-0 (#3435)
  Add error handling to contrainer removal (#3424)
  Prevent autofill prompt crash for edge case where a context menu is also visible on screen (#3417)

bwaresiak commented on Core/PersistentPixel.swift:
```diff
@@ -0,0 +1,299 @@
+            pixelFiring: Pixel.self,
+            dailyPixelStore: DailyPixel.storage,
+            onDailyComplete: { dailyError in
+                if dailyError != nil {
```
Shouldn't we ignore `Error.alreadyFired` ? There doesn't seem to be any tests for `fireDailyAndCount`

samsymons replied:
There are two tests already (`testWhenDailyAndCountPixels*`), but they miss this case - I forget that DailyPixel can error even in normal usage, good catch. Fixed this and added an additional test for it, thanks!

bwaresiak replied:
Right, I missed these when searching for them. :)

bwaresiak commented on Core/PersistentPixel.swift:
```diff
@@ -0,0 +1,299 @@
+                                additionalParameters: additionalParameters,
+                                includedParameters: includedParameters
+                            )
+                        ])
```
Validating assumptions here: typically daily/count combo is used for pixels that may fire more frequently and we want to understand affected devices - in case 'count' pixel floods us with data it can easily overflow the array (discarding other pixels) so we only get one signal.

I'm not saying we need to handle this now, but is seems like a hole in the current approach.

samsymons replied:
I've left this alone for the time being as I feel I'll still get some value from the current state, but I'd like to come back to it next week. At the same time, I'd also like to triage and add a storage error pixel, per a previous comment. Let me know if you feel strongly about addressing this one now before it's merged though.

bwaresiak commented on Core/PersistentPixel.swift:
```diff
@@ -0,0 +1,299 @@
+                                  error: Swift.Error? = nil,
+                                  withAdditionalParameters additionalParameters: [String: String],
+                                  includedParameters: [Pixel.QueryParameters] = [.appVersion],
+                                  completion: @escaping ([Error]) -> Void = { _ in }) {
```
Maybe we should use tuple and named parameters here?

bwaresiak commented on Core/PersistentPixelStoring.swift:
```diff
@@ -0,0 +1,176 @@
+
+    func append(pixels newPixels: [PersistentPixelMetadata]) throws {
+        try fileAccessQueue.sync {
+            var pixels = try self.readStoredPixelDataFromFileSystem()
```
I'd consider caching the state in memory so we can limit the number of reads.

bwaresiak commented on Core/PersistentPixel.swift:
```diff
@@ -0,0 +1,299 @@
+        for pixelMetadata in queuedPixels {
+            if let originalSendDateString = pixelMetadata.timestamp,
+               let originalSendDate = dateFormatter.date(from: originalSendDateString),
+               let date28DaysAgo = calendar.date(byAdding: .day, value: -28, to: currentDate) {
```
I think `date28DaysAgo` can be computed once outside of the loop.

bwaresiak commented on Core/PersistentPixelStoring.swift:
```diff
@@ -0,0 +1,176 @@
+        switch pixelType {
+        case .daily: return eventName + "_d"
+        case .count: return eventName + "_c"
+        case .regular: return eventName
```
I recall we wanted to move away from _d and _c - having this hardcoded here will make it harder to do so. Can we just store name along with suffix in `eventName`? I'm not sure there is any real value of storing if given pixel was daily or count other than name handling.

samsymons replied:
The `DailyPixel` API has been updated in the following way:
1. It now accepts a tuple of suffixes when sending a pixel
2. Any code using the `fireDailyAndCount` API currently has been updated to use the legacy suffixes
3. Finally, the `fireDailyAndCount` API was given new `_daily` and `_count` suffixes by default, so that no usage of the legacy suffixes gets added to the codebase going forward

The `PersistentPixel` class will now use whatever suffixes are given to it, and stores the final pixel name (including suffix) to disk, so the `PersistentPixelMetadata` struct no longer cares about which type of pixel it's dealing with.

Because of step 2 the PR has grown in size a bit, but for testing it you can remove the default value from `DailyPixel` and confirm that the app still compiles, as all uses have been accounted for and no pixels have been accidentally updated to use the new suffixes.

samsymons committed:
Merge branch 'main' into sam/persistent-pixels

* main:
  Add Events Firing for Phishing Detection Settings: Point to BSK (#3423)
  DuckPlayer: Temporary Fix for Watch In Youtube (#3437)
  Add 'Open in New Tab' support for DuckPlayer (#3431)
  update BSK dependency (#3434)

samsymons committed:
Only compute `date28DaysAgo` once.

samsymons committed:
Add a test for the DailyPixel `alreadyFired` case.

samsymons committed:
Change PersistentPixel errors array to tuple.

samsymons committed:
Merge branch 'main' into sam/persistent-pixels

* main:
  Remove `SubscriptionFeatureAvailability` from `AppDependencyProvider` (#3447)
  Release 7.141.0-2 (#3451)
  Do not notify the FE on experiment activation (#3450)
  point to bsk branch (#3444)
  bump bsk for content blocker rules fix (#3445)
  speculative fix for set bars visibility crashes (#3442)
  Release 7.141.0-1 (#3443)
  Fix browsing menu bottom offset when bar location set to bottom (#3440)
  Properly handle responses that should trigger download action (#3407)

samsymons committed:
Update DailyPixel to accept suffixes as a parameter.

samsymons committed:
Remove unused stub entry.

samsymons commented on DuckDuckGoTests/DailyPixelTests.swift:
```diff
@@ -349,19 +354,14 @@ final class DailyPixelTests: XCTestCase {
+    func testThatDailyPixelWithLegacyPixelSuffixAndCountWillAppendDAndC() {
         let expectation = XCTestExpectation()
 
-        stub { request in
-            request.url?.absoluteString.contains(Pixel.Event.forgetAllPressedBrowsing.name + "_d") == true
-        } response: { _ in
-            return HTTPStubsResponse(data: Data(), statusCode: 200, headers: nil)
-        }
```
The stubs in this test file weren't actually being used this entire time, since each test is using a mock pixel sender – as a result, these tests were not really testing anything. I've changed them to instead look at the pixel names sent to the mock.

samsymons committed:
Cache pixel metadata for read/write calls.

These base functions are not themselves thread-safe, but are always expected to run on the `fileAccessQueue`. A dispatch precondition has been added to ensure this. The reason for not making the base functions thread-safe is that some operations expect to call both of them in a single operation.

bwaresiak commented on Core/PersistentPixel.swift:
```diff
@@ -0,0 +1,292 @@
+        )
+
+        dispatchGroup.notify(queue: .global()) {
+            let errors = [dailyPixelStorageError, countPixelStorageError].compactMap { $0 }
```
Some leftover? Weird that linter didn't fail...

samsymons replied:
Huh, this is surprising. Let me fix it - thanks!

bwaresiak reviewed (approved):
Great job Sam, left one more comment around some warnings, but overall it's approved :)

samsymons committed:
Merge branch 'main' into sam/persistent-pixels

* main:
  Remove `voiceSearchHelper` from `AppDependencyProvider` (#3452)
  Update AutoClearSettingsViewController to use DI for app settings (#3448)
  Bump BSK (#3441)

samsymons committed:
Remove unused variable.
